_articles/scripting/Typescript/typescript-modifier.md

```diff
@@ -0,0 +1,231 @@
+@registerModifier()
+export class modifier_typescript_ancient_seal extends BaseModifier {
+
+}
+```
+
+As you can see, it's very similar to an ability, replacing `@registerAbility()` with `@registerModifier()`, and the `BaseAbility` extension with `BaseModifier`.
+Note that `@registerModifier()` takes care of LinkLuaModifier for you, so you don't need to call it on TS modifiers.
+
+### Typechecking Modifier Calls
+
+Before we continue, one thing we can do to add some more type checking is to link the ability to the modifier. Instead of relying on a string for the naming of the modifier, we'll instead link the class name.
+
+To do so, we simply remove the quotation marks around the modifier name, then add `.name` to it. See below the code before and after linking the class:
+
+<MultiCodeBlock titles="Before|After">
+
+```ts
+// Add modifier
+target.AddNewModifier(this.GetCaster(), this, "modifier_typescript_ancient_seal", { duration: seal_duration });
+```
+
+```ts
+// Add modifier
+target.AddNewModifier(this.GetCaster(), this, modifier_typescript_ancient_seal.name, { duration: seal_duration });
+```
+
+</MultiCodeBlock>
+
+This results at the exact name of the modifier as a string, which is enforced by Typescript.
+
+:::note
+If your modifier is in another file, you'll have to import it first before you can link it in the above fashion.
+:::
+
+### Coding The Modifier
+
+Alright. Let's set and apply some properties for the modifier, such as the parent, the ability and the particle effect. In addition, let's set some useful properties via modifier functions.
+Also, this is my personal choice, but I usually put ability specials as a class property so they can be easily used everywhere in the modifier.
+
+```ts
+@registerModifier()
+export class modifier_typescript_ancient_seal extends BaseModifier {
+	ability = this.GetAbility()!;
+	parent = this.GetParent();
+	particle_seal = "particles/units/heroes/hero_skywrath_mage/skywrath_mage_ancient_seal_debuff.vpcf";
+	resist_debuff?: number;
+
+	IsHidden() {
+		return false;
+	}
+
+	IsDebuff() {
+		return true;
+	}
+
+	IsPurgable() {
+		return true;
+	}
+
+	OnCreated() {
+		this.resist_debuff = this.ability.GetSpecialValueFor("resist_debuff");
+
+		// Add particle effect
+		const particle = ParticleManager.CreateParticle(this.particle_seal, ParticleAttachment.OVERHEAD_FOLLOW, this.parent);
+		ParticleManager.SetParticleControlEnt(particle, 1, this.parent, ParticleAttachment.ABSORIGIN_FOLLOW, "hitloc", this.parent.GetAbsOrigin(), true);
+		this.AddParticle(particle, false, false, -1, false, true);
+	}
+}
+```
+
+Okay, so the modifier is defined, but its main properties are not being applied: the silence and the magic resistance reduction, so let's do those next.
+
+### States
+
+The `CheckState` function that modifiers have is called every frame and sets the state of your modifier. The function gets a bunch of states and pairs each of them with a boolean that decides whether the state should be applied.
```

Comment:
This is not correct, it sets the state of the modifier parent, not of the modifier itself. It might be good to explain `parent`, `caster`, `GetParent()` `GetCaster()` and `GetAbility()` somewhere actually/

Reply:
Changed the explanation of the modifier states.

I decided to remove the ability and the parent and instead use GetAbility() and GetParent(). 
I also decided to not talk about those, because at the beginning of the ability for Typescript portion, I state that I expect you to understand how they work by going over Elfansoer's ability/modifier guides. Since I'm only focusing on the TS parts of them, I do not explain any of them, because otherwise, I'll also have to explain the particle attachment, which is already straying too far.